### ISSUE TITLE
Plasmoid bugged in Plasma 6

### ISSUE THREAD

**Issue by OlegAckbar** [Martchus/syncthingtray#239]
**Relevant components**
* [ ] Standalone tray application (based on Qt Widgets)
* [x] Plasmoid/applet for Plasma desktop
* [ ] Dolphin integration
* [ ] Command line tool (`syncthingctl`)
* [ ] Integrated Syncthing instance (`libsyncthing`)
* [ ] Backend libraries

**Environment and versions**
* Versions of `syncthingtray`, `qtutilities` and `c++utilities`: syncthingtray-qt6 1.5.0-1, qtutilities-qt6 6.13.5-1, c++utilities 5.24.7-1
* Qt version: 6.6.2
* Operating system (name and version): Arch Linux

**Bug description**
Plasmoid bugged in Plasma 6. It shows 2 entries in tray settings: one of them has no name and its setting can't be changed, the other one has no icon and its setting doesn't do anything for actual the tray icon of plasmoid (for example I have "always hide" option but plasmoid still stays in tray, see screenshots)

**Screenshots**
![Снимок экрана_20240314_233410](https://github.com/Martchus/syncthingtray/assets/9915567/b78d3b8a-d1b8-4a94-a665-372d570835af)
![Снимок экрана_20240314_233703](https://github.com/Martchus/syncthingtray/assets/9915567/1775d659-eb6d-45b7-bdd2-14b7061654fb)

**Comment by Martchus** [comment]
That's strange and something I haven't seen before. Not sure whether I can help here. It looks like some general setup problem. Maybe there's a helpful clue in stdout/stderr of the plasmashell process. You can open a terminal, kill and relaunch plasmashell from there and look for related log messages.

**Comment by Martchus** [comment]
But also check the README section about configuring the Plasmoid in general. It might help you understanding where the tray icon you see actually comes from.

**Comment by gudvinr** [comment]
I do have the same issue.

That could be caused by old (qt5) plasmoid not being pruned correctly on uninstall.

**Comment by Martchus** [comment]
I just had a look on my Tumbleweed system (where I haven't already migrated on a pre-release of KDE 6) and there it looks similar as in your screenshot (but not the same). So even after uninstalling the Qt 5 based version¹ the old entries stays in a disabled way. However, in my case the new entry looks fine (icon is displayed) and behaves like it should. So really only the leftover entry is the problem here.

If that's just a stale entry, allowing to delete it is possibly out of scope here. That's because the code in Plasma's system tray Plasmoid is responsible for managing these entries. However, I couldn't find were this stale entry would be configured so I'm not sure whether it is actually a stale entry. Maybe there's also something wrong with my metadata file causing this entry (in which case it wouldn't be a stale entry but a weird duplicate).

---

¹ On Arch Linux you could even keep the `syncthingtray` package installed; its current version no longer contains the Plasmoid anyway.

**Comment by gudvinr** [comment]
I removed all entries related to both plasmoids from `.config/plasma-org.kde.plasma.desktop-appletsrc` and `.config/kservicemenurc`, deleted `syncthingtray-qt6` package and restarted system. Now I have proper entry:  
![image](https://github.com/Martchus/syncthingtray/assets/1712806/e26b2857-54b4-4647-99bc-ad2528086816)

Now `.config/plasma-org.kde.plasma.desktop-appletsrc` has these:
``` 619 extraItems=<...>,Syncthing Tray (Qt 6)
620 hiddenItems=<...>,Syncthing Tray (Qt 6)
```

It's all weird AF. Now it stores name instead of ID (which is very questionable given that multiple applets might have same human-readable title)

**Comment by Martchus** [comment]
Both cases are quite weird. I'll try to see how it behaves in my case later.

@OlegAckbar Is the Plasmoid generally working for you (so this is *just* about the entry like in @gudvinr's case)? And have you tried what I suggested in my first replies?

**Comment by OlegAckbar** [comment]
@Martchus Yes, plasmoid works just fine, no problem with that. Only I can't hide it in tray.
![изображение](https://github.com/Martchus/syncthingtray/assets/9915567/19d3855d-1ec7-409b-85ff-c0f8db919fc7)

**Comment by OlegAckbar** [comment]
@Martchus I've tried launching plasmashell from terminal but I didn't see any weird warnings or errors mentioning syncthingtray (only those mentioning that my locale is not supported).

**Comment by OlegAckbar** [comment]
[systemlog.txt](https://github.com/Martchus/syncthingtray/files/14618547/systemlog.txt)
here's journalctl log filtering plasmashell. Maybe you'll find something interesting

**Comment by OlegAckbar** [comment]
I've built syncthingtray-qt6 on my Arch VM with testing repositories enabled and have exact same issue.

**Comment by Martchus** [comment]
I also couldn't find anything useful in those logs after a brief look.

I did a few more tests on my system and things are even weirder: First of all I should note that I have multiple screens and thus multiple panels and thus also multiple system tray plasmoids. Plasmoid instances are configured separately and the system tray plasmoid is no exception. So I had two relevant sections in `~/.config/plasma-org.kde.plasma.desktop-appletsrc`. I don't think any of the sections (and also not any file in the whole `~/.config` and `~/.local/share` directories) contained any references to the old plasmoid, though.

So I stopped plasmashell, deleted all keys in the relevant sections of one of the system tray plasmoids and then in fact one Syncthing Tray icon was gone and the other one preserved. When trying to re-configure the gone Syncthing Tray icon via the system tray plasmoid settings I noticed that the weird entry was gone. Only the correct entry was shown and everything worked as expected. Then I opened the settings of the other system tray plasmoid and there was the weird dysfunctional entry again.

So I stopped plasmashell again and then deleted all keys in the relevant sections of all system tray plasmoids. Then all Syncthing Tray icons were gone. When configuring them back everything looked good at first but when opening the plasmoid settings another time the dysfunctional entry was back.

So this is not a stale entry of the old version of the plasmooid. It gets newly created for the current version of the plasmoid under some circumstances which might involve having more than one system tray plasmoid.

I must also note that plasmashell crashed when applying changes of one system tray plasmoid while having the settings window of another one open.

Somehow I think all of this has not much to do with Syncthing Tray but it is a thing of how the official system tray plasmoid handles its settings/entries.

**Comment by gudvinr** [comment]
Nope, apparently it isn't magically fixed and I am having same issue again

**Comment by Martchus** [comment]
Yes, as I said I could reproduce it. But it is complicated and fixing it probably requires collaboration with Plasma upstream. So far I haven't spent the effort considering I'm not really bothered by this additional invalid entry. (In my tests the real entry always worked so the Plasmoid is fully usable.)

**Comment by gudvinr** [comment]
I have only one tray applet and only single plasmoid instance, FWIW

**Comment by LinAGKar** [comment]
I only have one entry in the list:
 
![Screenshot_20240404_113159](https://github.com/Martchus/syncthingtray/assets/3390621/80f92b0f-7821-466f-9ae8-a2740a7dcac2)

and get no syncthing tray icon at all. If I try to add syncthing as a plasmoid it just vanishes again when I restart plasmashell.

**Comment by LinAGKar** [comment]
It doesn't happen to any other plasmoids though, only Syncthing disappears. I'm also still on 1.5.0.

For the tray icon, apparently it does show up if I change the visibility to something else instead of the default "Enabled" (maybe that's a mistranslation). And then the duplicate row also shows up for me, and it won't go into the overflow menu even if I set it to always hidden.

**Comment by gudvinr** [comment]
Filed an issue in KDE bugtracker:  
https://bugs.kde.org/show_bug.cgi?id=485072

**Comment by Martchus** [comment]
Still nothing that I think I can fix from my side. If this is important to you, you'd probably have to dig into the code yourself.

**Comment by LinAGKar** [comment]
I think it's not right to close it while the issue still happens

**Comment by livexia** [comment]
with plasmoid in plasma 6 after adding it works normal, but after reboot the plasmoid disapper.

**Comment by Martchus** [comment]
I don't think anyone can help you with so little information. I suggest you have a look at https://github.com/Martchus/syncthingtray?tab=readme-ov-file#configuring-plasmoid and further documentation sections linked from there.

**Comment by livexia** [comment]
>1. It can be shown as part of the system tray Plasmoid. 

With this method somehow syncthing disappears, after I remove syncthing from `.config/plasma-org.kde.plasma.desktop-appletsrc` and reinstall syncthingplasmoid-qt6. This method works again.

>2. It can be added to a panel or the desktop like any other Plasmoid.

With this method, panel disappear after reboot, but can be added again.

I think this program is great thank you.

**Comment by Martchus** [comment]
So at least the 1st method works. I have no idea why the 2nd method is not persistent for you. I suppose this has more something to do with the general configuration mechanism of the Plasma shell and less with my specific Plasmoid.

**Comment by olifre** [comment]
As a potential extra data point, I can add that I have three Gentoo systems with exactly the same package versions, but different evolution, and two systems show the tray icon while one does not, i.e. two systems behave as the original reporter's system and one as @LinAGKar . 
It "smells" like this is something fuinny like loading order / iteration order being based on the number of the inode or some other number which can be different in each installation of the very same OS. 
I could sadly not find an obvious difference between the systems :disappointed: .

**Comment by xelra** [comment]
So I don't really have a solution either. When I add the Plasmoid by itself, it doesn't get added to `plasma-org.kde.plasma.desktop-appletsrc` and hence doesn't persist over a reboot. So there seems to be some issue there.

What does work though is adding it to the System Tray group of items, by toggling its visibility on in that group. Then I can also find it in `plasma-org.kde.plasma.desktop-appletsrc` as part of `extraItems` and `shownItems`.

At least this is a workaround for getting the icon in the tray for now.

**Comment by olifre** [comment]
> What does work though is adding it to the System Tray group of items, by toggling its visibility on in that group. Then I can also find it in `plasma-org.kde.plasma.desktop-appletsrc` as part of `extraItems` and `shownItems`.

This is the approach I am using on the three machines mentioned in my previous comment. However, they still behave differently, i.e. two systems show the system tray icon and one does not.

**Comment by olifre** [comment]
Still present with Plasma 6.3.4 and Qt 6.8.3. It really smells like an upstream bug, but sadly there has not been real activity on the KDE bugtracker (apart from checking that it's still broken) and it's strange that other Plasmoids do not seem affected.

**Comment by olifre** [comment]
Still present with Plasma 6.3.5 and Qt 6.9.1 on my systems.

**Comment by gudvinr** [comment]
Same with Plasma 6.4.3 as well.

**Comment by benhaube** [comment]
I am still seeing this issue on Plasma 6.4.4 and syncthingplasmoid-qt6 2.0. I tried the fix/workaround mentioned above, but it didn't work for me. I still have the invalid "Disabled" entry at the top of the list, and changing the visibility of the entry listed lower down the list does nothing.

**Comment by gudvinr** [comment]
upstream (KDE) closed [an issue](https://bugs.kde.org/show_bug.cgi?id=485072) as 
> it's now an issue in the 3rd-party Plasmoid code itself

**Comment by Martchus** [comment]
I still haven't learned anything new that would help me finding out what's wrong with the Plasmoid code, though.

**Comment by Martchus** [comment]
This issue is probably fixed with the commit auto-referenced by GitHub. I wasn't really able to reproduce it anymore in my test env before anyway so I'm not quite sure whether that commit will also fix this issue. (It definitely fixed https://github.com/Martchus/syncthingtray/issues/339, though - and both issues have probably the same cause.)

**Comment by Martchus** [comment]
The commit that fixed #339 is now part of the latest release and I also haven't noticed the problem anymore on my setups.
